### PR TITLE
fix: Improve typing for CSV parser

### LINKED ISSUE #3266: Improve typing for CSV parser
(This  issue is part of https://github.com/denoland/deno_std/issues/2291#issuecomment-1425887042.)

Currently, I think there are some things that could be improved in the type definition of the parse function in the csv module.

1. Not all candidates are included in the list of optional argument completions.
I should also see `columns` and `skipFirstRow` in the completion suggestions, but they are not.
image

2. The return type of the function below should be `string[][] | Record<string, unknown>[]`, but it is `string[][]`.
image
The overload below doesn't seem to be utilized.
https://github.com/denoland/deno_std/blob/cf14c9b21234f8d99ac1ae819d912e98de47be9c/csv/parse.ts#L345-L350

3. How about setting the return value to `Record<string, string|undefined>` instead of `Record<string, unknown>` when passing the `columns` option or `skipFirstRow` option? for example:

    ```ts
    parse("", { skipFirstRow: true }); //=> Record<string, string | undefined>[]
    parse("", { columns: ["aaa", "bbb"] }); //=> Record<"aaa" | "bbb", string>[]
    parse("", { columns: ["aaa", "bbb"] as string[] }); //=> Record<string, string | undefined>[]
    ```
4. Passing a _readonly array_ to the `columns` option results in a strange return type of `string[][] & Record<string, unknown>[]`.

    image

I am working on this fix.

### PR DESCRIPTION
**Pull request by ayame113**
close #3266

Here's a [link to the playground](https://www.typescriptlang.org/play?ts=5.0.2#code/PTAEF5K6dAXAngBwKagCYoGYGcIwMKOMgChSUAPJAewCc5QBLAOzhTqwEMBjNAJRRd0AeSRwmNFngDepUKBwokXOlzj0A-AC5FcOqwDmAbnmgeNALaWUbHXoMsTZ-U0sAZIeiMBlFX3sAIxoaABshFlMFUK4ALwQARQBXGnYcIJDwrkizLCYUUPQcAAUOQQs6dHsWJMtAjlMAX3IqWgZmNg5uPlBi1SUxCSk8KnYWItBBYUHJaVA5BRwAayYkADEmOhw4fhoAdwywiKjzMNrpezovKVCEByMAbQBdJpbqekYsJJYeIZZQFRbFAAClYSCScF020chgANKAaOJ7N9MHkWCh0ABKKGuJzPZ6mVofUBfH5-AH9FAAHgs0kYABVQKNbBM+kCZsMAHzAsxgiE4mGwsyIyGgelC7G9SmCHBJUJwKlsgbiWY4eH0zmE97tUm-WYUoE04YMpmUMasykc6Tc3kscGi6FGIUKEW6RngeaKFbrTbbXZ7XTcUJKUCNUBcPDi0iSpUoGVyhWxq1qsWc+ZmK5wJJ0f7PcN4b5LFj7f4RqVA+PyxWWlXDdWa0jNN5tRiIVDlpSVxM1v4pjUQMwgUAASSwoAABhZQuccOOEbX-kw8DhUDwmHkMfC4AALNCZ7P-NtoJeTFAVdDwZAoAB0ZkZzPGeCTC7wADJPVOZ7orsIbndQSwWAcKAADCZyWHMD4TI6TiYnmzQKAominueADqTA7mB04QVSWEzpyzyDmAo4TssqwbFsOz7HOIr6ieK5nuu+ToFuu6gPuOaXu2J7lPQF5HreChuqa5pPj2qqgO+MheuRvpUQG8B0EkaAIYhyG8ZUVIwXC9xOKAAA+oAotgrAYgRTxESOY6TuB0hztkF7jmRPqUf6NELsweAAPrFnAXmKKuTGbvAbEcYeV6ebphj4k8gmgMJUFieyL6SR+tnpLoxlohixgyS5fr7PYQYhoZWWmReqlIVFMWWQUSisXuKBZpxR6RRp-ERfQ1VPM8cW6O1Wm4jp2kGUZ4wmei6DmaN2kxaYpBDu16GYbZVIAERcJta36WtgR7WtabgGmA0bVtO17YEa3wtpnILWAS0YdueE4TdEDHWefGDYKUWjWVk23a1D0rdhLC4elIkssuQ2HT9iWgelZjqR9mnaddQ2-eN2VTWY-XI+gYMgymN3zXd+AkPAKDbOYEZoCQdP02QpMkcW87kieKg4EoLEYNgXAJngGjdYRCynHSBpcxA4sghtAA6LCXZiJxDgoAB6miNqQIu0lTgISx6uvS1wcuXfCf0Yorlmq+rzQi0OTnehRBV7HOlhcHc9SKcpCJ0CSXDBjelk2aENC1LOoCu+7aBmxeXWzbFlkkQgIcYFIADkjBFvspq8HAtygHs27qCFaAAG5+17NBjjuaC0VInnOqAQ7V+xTUHlxx5QzCMWjQNqNRfp0fmWY2uMLX0i6M+vaSzIjQnCPUsXvrlLArL8tXazqoWwoyugGrGta8aC+SwbK9G2v8Iz1vjdgFbGuk85jvyRvUia8Ph8G4vUun8b6-SQ-cn+kypjcqoYr47z3jbN+YsP7H2XqvE2np-6uX2IGP2IZGhgJvrva2r8FDzxgUvIE39z6IIdgAlBnsVKYMQhA8gQ5PwQTwGPXBosdaUk-ifeBv9TiEyAaiEBGClZYNoQfaB7DYFEI2rCQIcseCwnQNwhh49QAPFOlwdeu19pPFAUImhODRFsKBBwuBZ8EHSSUTgXQqjNrqO0WWOOOjLbYLvkOJBTtn7-AANQ8JnB4lh+DxGEKUMQsxeVH6AN9v7eEFi+ETQxI47ewj9FQMMXrL+XCL5hPIQpfQylonpViVjBJ189H7xSYwAh6TTHcLcfJVBUSfGMKsWota2jBFOJEeUo+QTDY-0ybUiJuSUD5N4SolpbTqG32aEAA) to try out what the type definition looks like.

### CODE REVIEW

File: csv/parse.ts [hunk 1]
```diff
@@ -286,7 +287,8 @@ class Parser {
   }
 }
 
-export interface ParseOptions extends ReadOptions {
+export interface ParseOptions<Columns extends string = string>
+  extends ReadOptions {
   /**
```
Comment (ayame113) [inline csv/parse.ts:292]: Note: In order to achieve the following, I had to make the `ParseOptions` type generic and let TypeScript guess the concrete values for `columns`.

```ts
parse("", { columns: ["aaa", "bbb"] }); //=> Record<"aaa" | "bbb", string>[]
```

**Comment by ayame113** [comment]
Errors in CI will be fixed by #3280.

File: csv/parse_test.ts [hunk 2]
```diff
@@ -819,49 +819,90 @@ Deno.test({
...
+    // columns option
     {
       const parsed = parse("a\nb", { columns: undefined });
-      type _ = AssertTrue<Has<typeof parsed, string[][]>>;
+      type _ = AssertTrue<IsExact<typeof parsed, string[][]>>;
+    }
+    {
+      const parsed = parse("a,b\nc,d", { columns: ["aaa", "bbb"] });
+      type _ = AssertTrue<
+        IsExact<typeof parsed, Record<"aaa" | "bbb", string>[]>
```
Comment (kt3k) [inline csv/parse_test.ts:874]: This looks nice improvement. Thanks!

**Review by kt3k — APPROVED** [review]
LGTM

**Pull request merged**